### DISCUSSION TITLE
Recent UART Changes

### DISCUSSION THREAD

**Discussion by vhdlnerd** [stnolting/neorv32#537] [General]
I'm disappointed with the recent removal of the UART hardware handshaking lines.  This breaks all my neorv32 based designs. I'm a little shocked that functionality would be removed from a basic interface.  I use the UARTs at 921600 baud sending commands to the neorv32 that take time to process -- hardware handshake lines are required to avoid dropping commands.

The two workarounds mentioned in the change notes (adding a non-crippled UART as a custom function or a wishbone module) are equally poor. Either way means having a custom bootloader and worst not being able to use any of neov32's SW library functions that uses UARTs (neorv32_uart_printf, ...).

Brian

**Comment by stnolting** [comment]
Hey @vhdlnerd,

sorry to hear that the latest changes broke your design setup.

I tries to keep the peripherals as small and easy to understand as possible. I thought that the HW flow control was just another over-engineered feature that is rarely used. Furthermore, I thought that removing it would shrink the size of the UART. Apparently, however, I was too hasty and both assumptions turned out to be wrong. So, sorry, and thank you for the constructive feedback!

I think it should be no big deal to reintegrate RTS/CTS flow control. Still, it might be possible to shrink and simplify it a bit. Maybe a single control register bit would be sufficient to enable both handshake signals (if both are not required then they can be left unconnected / tied to zero)!? 🤔

I'll make a PR with a first proposal - would you be willing to help debugging (if required at all)? :wink:

By the way, I think this case is a nice example for a community-driven project: sometimes I do not see what the actual users are seeing/requiring. So, as always, feedback is highly welcome!

**Reply by vhdlnerd** [reply]
First after rereading my post, the tone was a little harsher than I intended -- I should not post when I'm in a hurry and tired. Sorry about that.

Sure, I can help test the RTS/CTS update you come up with, if needed.

**Reply by stnolting** [reply]
No worries! I use a lot of open source stuff and quite often I am like "What the #*!&?! Why did they change that???" - So I can understand you 😅 

> Sure, I can help test the RTS/CTS update you come up with, if needed.

Thank you very much! :+1:

**Comment by GideonZ** [comment]
Agreed. Even the basic UART should support RTS/CTS.
The cost of these signals is near zero, so I would suggest including them.

**Reply by stnolting** [reply]
Agreed! 👍

**Comment by vhdlnerd** [comment]
As to what my current neorv32 project is:
I'm using the Nexyx A7 100T board. I've implemented a DDR2 wishbone memory controller with a L2 cache. So, the neorv32 has a large 16MB RAM space to play with.
I also implmeneted a highly configurable (at build time) VGA text mode controller. Supports almost any resolution from 800x600 up to 1920x1080 and various fonts.
I plan to implement a keyboard controller at some point.

My main SW project I'm running on this setup is a Forth  language (interpreter) that I wrote almost completely in RISC-V assembly (yes, I'm that crazy 😃).  I need the HW handshaking lines for my Forth. I download Forth code via a simple ASCII upload over the UART.  Since the Forth code sent is complied on-the-fly, there is a delay before it can accept the next piece of code.  This will cause UART overruns without the handshaking lines.

I hope to have my neorv32 setup to a point to upload to github soon.  I want to link to it off the neorv32-setups repo if possible.

**Reply by stnolting** [reply]
Interesting project! Especially the L2 cache sound interesting (an internal data cache is on my to-do list for ages...).

> My main SW project I'm running on this setup is a Forth language (interpreter) that I wrote almost completely in RISC-V assembly

Impressive! :exploding_head:

> I hope to have my neorv32 setup to a point to upload to github soon. I want to link to it off the neorv32-setups repo if possible.

That would be really great! :+1:

**Comment by stnolting** [comment]
For the records:
I made a (first?) proposal for re-integrating the RTS/CTS flow-control in #541. I have done some simple testing and so far everything seems to work fine.

**Reply by vhdlnerd** [reply]
I tested the newly re-added RTS/CTS UART signals and everything worked in my projects.
Thanks for putting this feature back in!

**Reply by stnolting** [reply]
Thank you for the feedback!
